ipi/engine/normalmodes.py

```diff
@@ -730,15 +769,13 @@ def get_fspring(self):
         if len(self.bosons) == 0:
             return self.transform.nm2b(dstrip(self.fspringnm))
         elif len(self.bosons) is self.natoms:
-            return self.vspring_and_fspring_B[1]
+            return self.vspring_and_fspring_B[1].reshape((self.nbeads, 3 * self.natoms))
         else:
-            # raise("@NormalModes: Implementing mixtures of B and D")
-            f_distinguish = self.transform.nm2b(dstrip(self.fspringnm))
-            # zero force for bosons
-            for boson in self.bosons:
-                f_distinguish[:, 3 * boson : (3 * boson + 3)] = 0.0
+            f_distinguishable = self.transform.nm2b(dstrip(self.fspringnm))
+            f_all = f_distinguishable.reshape((self.nbeads, self.natoms, 3))
+            f_all[:, self.bosons, :] = self.vspring_and_fspring_B[1]
 
-            return f_distinguish + self.vspring_and_fspring_B[1]
+            return f_all.reshape((self.nbeads, 3 * self.natoms))
```

Comment:
All this above would be simpler if you computed it in a single f_and_v function.